### DISCUSSION TITLE
Thank You!

### DISCUSSION THREAD

**Discussion by jrft1379** [maforget/ComicRackCE#78] [General]
Greetings!

I've been using ComicRack for **_years_** and I thought there was no way it was ever going to be brought back. Simply put, no other comic program holds a candle to CR. Man, was I impressed and pleased to discover this!

I am **_not_** a coder or anything else: I am simply an end user. As such, I have no idea if things I've wished for over the years are even possible, but now there is at least somewhere to ask. I'm hopeful that there can finally be definitive answers to some of these. I've gone over all the back documentation here and on Reddit, so I apologize in advance if something has already been addressed.

Just a few quick notes about how I have my specific UI laid out: I view everything in thumbnail mode. I have my group, stack, and sort/arranged all set to Alternate Series (which I have meticulously maintained since it seemed to be the only way I could organize my books exactly as I wanted them). Within my stacks, I have my books sorted by format then publication date.

Here are a few things I've noticed over the years:

- As my collection has grown (up to 85k), the viewer has had trouble maintaining a specific book as the top of the stack. Simply put, If I start assigning issues as top in the As, by the time I get to the Fs, the library forgets what the top book was back in the As. I'm assuming that this is because I have such a large library. The same is true when I tried to set custom thumbnails.
- I'm not sure if it is because of the size of my collection or because it can't default to a set image each time, but I've always had a delay when I close a stack or switch from a list to my library as it creates thumbnails for the top of the stacks.
- Again, I'm sure it's because of my library size, but if I change views too quickly (like reverse the alphabetical ordering), It can sometimes freeze the program for up to 3 minutes (I've timed it before). It will eventually recover and has never crashed.
- If I am in a library stack, I switch to another list, and then go back to the library the program seems to forget all the books that are in the stack unless I close the stack and reopen it.
- The reason I use the Alternate Series field to group, stack, and sort is because there is no "Sort Name" and "Display Name" for individual files. I love this function in apps like Plex, but I have no idea if it is practical in a program that has such flexibility in it's organization.
- I love the "Show Group Headers" function, because it allows for quicker navigation. But, it only collapses down so far, and I wish it could be made narrower than it currently allows.
- Although I've come to accept it, I get very annoyed at the "Annual" format, and that it defaults to showing that. Annuals are not a format! (Yet surprisingly some actual formats like "Treasury" and "Digest" are not automatic options).
- I really wish I could make the default for "Proposed Values" no.
- Will the recognized publishers and imprints be updated to add in new(er) ones?
- Also, what about sub-imprints (like Wildstorm: Cliffhanger, or Skybound: Energon Universe)?
- Will the various publisher logos and imprint logos be updated eventually? I can actually help with that if I'm given dimensions they need to conform to.

Again, thank you to everyone involved for bringing this amazing program back to life!

**Comment by maforget** [comment]
First let say, that I don't know all the code base already and still figuring out some parts of the program. It is a big code base.

> As my collection has grown (up to 85k), the viewer has had trouble maintaining a specific book as the top of the stack. Simply put, If I start assigning issues as top in the As, by the time I get to the Fs, the library forgets what the top book was back in the As. I'm assuming that this is because I have such a large library. The same is true when I tried to set custom thumbnails.

I am not aware of the mechanism with stacks completely yet, but I don't see that there would be a limit and it would "forget". Maybe closing the program to force a save of the DB would help. This would need specific steps to reproduce to be able to try to figure it out. Looking at the DB, it seems that for each list there is a StackConfig that hold the Configs based on the Stack name. Maybe you have so much stacks that multiple with the same name appear in the same list (or library in your case).

> I'm not sure if it is because of the size of my collection or because it can't default to a set image each time, but I've always had a delay when I close a stack or switch from a list to my library as it creates thumbnails for the top of the stacks.

Thumbnails are cached at the beginning if you let it create them it will get better and there shouldn't have need to recreate them always. But keep in mind that creating thumbnails needs to access the file, so if your file are on a slow storage like a NAS and you open a list containing 85k item it will take a while to access each and every file to create the thumbnails. That being said I haven't experienced a noticeable delay opening and closing stacks.

> Again, I'm sure it's because of my library size, but if I change views too quickly (like reverse the alphabetical ordering), It can sometimes freeze the program for up to 3 minutes (I've timed it before). It will eventually recover and has never crashed.

Again never experienced this, except with some new property I've added that would check the actual file format and not just the extension. It would freeze until it was accessing everything. But normally CR is good at doing things like this in the Background and not freeze the UI. But Sorting can an almost exponential operation that with 85k will take a lot longer. 

I would suggests that you create sub-folders to reduce the count of books, instead of just using the library that shows everything. It may fix some of your issues. Honestly the only time I use the library is when searching for something I don't know  where it is located, because usually I have plenty of lists and folders for each language, publisher, etc.

> If I am in a library stack, I switch to another list, and then go back to the library the program seems to forget all the books that are in the stack unless I close the stack and reopen it.

That was reported already #68. I haven't figured out how to reproduce it enough to test it.

> The reason I use the Alternate Series field to group, stack, and sort is because there is no "Sort Name" and "Display Name" for individual files. I love this function in apps like Plex, but I have no idea if it is practical in a program that has such flexibility in it's organization.

What exactly does your Alternate Series field does than the Series doesn't? You can tell the program the Articles to ignore when sorting like "The". This can be modified by editing the `ComicRack.ini` file. If the sorting is unrelated than using other fields like that is the way to go. You could use Custom fields and set value like that and sort and group with that also. Using plugins like Data Manager to automate the work. I've added a Virtual Tags, that is limited to putting only multiple fields together, but if it gets functions like substring & RegexReplace and things like that it could be very powerful.

I feel like we have multiple way to deal with the problem and adding an additional field just to sort series field would more annoying to maintain than anything else. Imagine having to update 85k books because you renamed the Series to update the Sort Series also. 

> I love the "Show Group Headers" function, because it allows for quicker navigation. But, it only collapses down so far, and I wish it could be made narrower than it currently allows.

This I don't understand you can just drag the "Show Group Headers" window to the right to reduce it's width. There is a minimum width byut it is very small.

![image](https://github.com/maforget/ComicRackCE/assets/11904426/2e337e88-f6f3-487d-8acf-ed80d9b4397e)

> Although I've come to accept it, I get very annoyed at the "Annual" format, and that it defaults to showing that. Annuals are not a format! (Yet surprisingly some actual formats like "Treasury" and "Digest" are not automatic options).

Well Annual is useful, if it annoys you, you can edit the DefaultLists.txt in the install dir. Problem is that it will be recreated each time you update. You can copy that file in your Data dir to update it, but it is more useful to add things than remove them. The fact that it is there doesn't impact anything if you don't use it.

> I really wish I could make the default for "Proposed Values" no.

I understand, but we could either have it to no by default, then there would be no value detected a new book would be empty without any values. Or we could commit them automatically, but then it would be harder to determine which file as auto detected values and which aren't. You can easily automate it with DataManager by just setting `EnabledProposed` to False. It does the same than changing it to No manually. Or use the included script to commit values.

> Will the recognized publishers and imprints be updated to add in new(er) ones?
> Also, what about sub-imprints (like Wildstorm: Cliffhanger, or Skybound: Energon Universe)?

There is absolutely no recognized publishers in ComicRack all the data that is available is done via user provided values. If a value exists in your DB it will appear has a choice. Logo aren't the same.

> Will the various publisher logos and imprint logos be updated eventually? I can actually help with that if I'm given dimensions they need to conform to.

They already have, check this issue if you want to participate. #62. The size seems to be 64x64, but there is no limit and i've discovered that 64x64 seems comes out very blurry, so I've let the new ones I added in there full resolution. Just use transparent PNG if possible.

**Reply by jrft1379** [reply]
Hey maforget!

Thanks for the response. I 100% understand that this is a tremendous undertaking, so I'm more than happy to be patient as y'all learn what is going on with the code. Take your time, please.

**Stack Assigned Tops/Custom Thumbnails:** I'm really not sure why it happens either. It has happened on multiple occasions, so I know it wasn't a one-time fluke. I don't believe the stack names are the same in StackConfig, as I specifically arrange my stacks via the Alt Series field, and that is unique for each stack. I can reproduce and document the steps, but it will be a little bit, since I need to go through so many stacks to cause it to happen.

**Stack Thumbnails:** I guess I misstated that. Sorry. It isn't that there is a delay in the opening and closing of the stacks, but a delay in the thumbnails populating once I leave the stack into the library of list. I'm guessing this has to do with the size of my collection and that it won't seem to accept assigned top issues for every stack and just randomly pulls an issue to have represent the stack.

**Freezing:** Yeah, I figured it was do to the size of my library. I do indeed have plenty of lists, smart lists, folders, and sub-folders to try and break my collection down into manageable chunks. Sometimes that helps, sometimes not.

**Forgetting Books in Stacks:** Cool, glad it isn't just me. I'm sure it will be parsed out at some point.

**Alternate Series:** So the reason I used the Alternate Series field is because of how I like my comics to be sorted vs how I want the issue name to be displayed. For example, the second volume of Firestorm: during the run of that series it went from _The Fury of Firestorm_ to _Firestorm: The Nuclear Man_ to _Firestorm_. I want the true title reflected in the series name of each issue, but I want all of them in the same stack. There are other reasons, too: I like to stack events with one-shots together (think _Convergence_ or _Future State_); I like to separate out the various series via volume (yes, I know you can have more than one selection to stack).

So basically, I transfer the scraped data that goes into the Alternate Series field - generally what I would consider a Story Arc - to that field and put my own value in the Alternate Name field. I know this may not be how others would handle this, but it works for me. And I get that adding extra fields would be a big ask. So no worries on that.

**Group Headers:** Yep, I have the group header just like you have it in the pic. But I personally don't think it is narrow enough. It would be great if there was no limitation so that each user can decide what works for them.

**Formats:** You are correct that formats I don't use don't really do any harm when they are not used. They're just superfluous. However, combining that with the fact that the field is considered a Proposed Value - which is automatically set to yes - I'd just as soon not have the format there unless I set it. I have zero interest in going in each time the program updates and editing the txt file.

*Proposed Values:** Again, I'm sorry for not being clear. My suggestion is that each user have the option to make the default "yes" or "no" based on their personal preferences. That said, I will explore your suggestion with the DataManager.

**Publishers & Imprints:** I see what you are saying. That makes sense.

**Logos:** I will check out #62 and see how I can help. I'm a graphic designer by trade, so making logos into transparent PNGs is easily done.

Thank you again for your response and taking up the challenge of leading this effort!

**Reply by maforget** [reply]
- Regarding the proposed values, I understood your point. What I am saying is that when you change it to No, any value that are in pale grey would be lost. So by setting the default to no, the book would just detect nothing, making scraping after that difficult, because you don't even have a series to being with. The only alternative is to commit these default value so they are taken has a fact and not just proposed. It's doable, but I don't know the impact on every user. Some don't understand the idea behind proposed values and would just turn them off, but would loose the ability to detect the places that have not been set.

- Regarding your Firestorm problem. One solution is also to use either the SeriesGroup or StoryArc field. I use the SeriesGroup to group say Spider-Man and with the new Virtual Tags, I've created a tag that shows SeriesGroup - Series, so I can group them like that and see all the different Spider-Man series next to each other, regardless of their name.

    - Another way would be to change the Series to FireStorm manually and be done with it. I am guessing the times that the Series changes name mid run is mostly the exception than the rule. It would be easier than to change all the Alternate Series. Again DC & Marvel have done some real head scratcher to deal with, like renumbering. We all deal with them in a different way, adding extra fields is very easy, updating the UI more annoying, but with all the tool we have like custom and reusing a field that you aren't using for something completely different isn't illegal.

     - One thing that is very helpful in these case is to use Data Manager to fix these issues. For example, ComicVine sometimes has TPB, has there own volume. So I did a rule that would check for Title that have Volume # and a Number 1 and extract the Number from the title into the Number. The series are also usually are also separated by a :, the series to the left and the title to the right. So I did a rule to split them and assign the right part into the Title field. It's pretty much automated and works on 95% of times.

**Comment by jrft1379** [comment]
Just wanted to post to let you know I upgraded to a new system (is 12 years too long to have a desktop going non-stop?) and some of the performance issues have solved themselves. I'm not having nearly the delays when switching between lists, stacks, and the library anymore. 

I've not tried to see how the new system handles set "top of the stack" graphics, though. I have no clue if it can handle more than approximately 1,000 like it did with the old system. I do wonder if it's a limit of the program or the system.

**Reply by maforget** [reply]
I did a fix this week, that speeds up the program when you have a lot of smart lists. It would take over 5 sec opening the smart list edit window (on my system), it went down to around half a second. That changed also helped searching & dragging smart lists.

12 years is a lot especially with spinning hard drives.
